File: server/extract.js
import path from "node:path";
import mammoth from "mammoth";
import pdfParse from "pdf-parse";

const TEXT_EXTENSIONS = new Set([".txt", ".md", ".markdown", ".rtf"]);

export async function extractManuscriptText(file) {
  const ext = path.extname(file.originalname).toLowerCase();

  if (ext === ".docx") {
    const result = await mammoth.extractRawText({ buffer: file.buffer });
    return normalizeText(result.value);
  }

  if (ext === ".pdf") {
    const result = await pdfParse(file.buffer);
    return normalizeText(result.text);
  }

  if (TEXT_EXTENSIONS.has(ext)) {
    return normalizeText(file.buffer.toString("utf8"));
  }

  throw new Error("Unsupported file type. Upload a PDF, DOCX, TXT, or Markdown manuscript.");
}

function normalizeText(text) {
  return text
    .replace(/\r\n/g, "\n")
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{4,}/g, "\n\n\n")
    .trim();
}

File: review_it/extract.py
import re
from io import BytesIO
from pathlib import Path

from docx import Document
from pypdf import PdfReader


TEXT_EXTENSIONS = {".txt", ".md", ".markdown", ".rtf"}


def extract_manuscript_text(filename: str, content: bytes) -> str:
    ext = Path(filename).suffix.lower()

    if ext == ".docx":
        document = Document(BytesIO(content))
        text = "\n".join(paragraph.text for paragraph in document.paragraphs)
        return normalize_text(text)

    if ext == ".pdf":
        reader = PdfReader(BytesIO(content))
        text = "\n".join(page.extract_text() or "" for page in reader.pages)
        return normalize_text(text)

    if ext in TEXT_EXTENSIONS:
        return normalize_text(content.decode("utf-8", errors="replace"))

    raise ValueError("Unsupported file type. Upload a PDF, DOCX, TXT, or Markdown manuscript.")


def normalize_text(text: str) -> str:
    text = text.replace("\r\n", "\n")
    text = re.sub(r"[ \t]+\n", "\n", text)
    text = re.sub(r"\n{4,}", "\n\n\n", text)
    return text.strip()
